### DISCUSSION TITLE
Question on the KKS AC kernels

### DISCUSSION THREAD

**Discussion by xueyang94** [idaholab/moose#18329] [Q&A Modules: Phase field] [answered]
Hi everyone,

There might be something really simple that I misunderstood, but why both the KKSACBulkC and KKSACBulkF kernels do not have the mobility L multiplied to their residual and on-diagonal Jacobian? Only the off-diagonal Jacobian is multiplied by L. Another kernel for the order parameter, ACInterface however, has the L in its residual and Jacobian functions. Thanks.

**Comment by GiudGiud** [comment]
@jiangwen84 if you could please help with this question

**Comment by dschwen** [comment, marked as answer]
The non split KKS formulation is not recommended for use. Please use the split formulation. The non split has faulty Jacobians.

**Reply by dschwen** [reply]
Source: I wrote all those classes

**Reply by xueyang94** [reply]
@dschwen Yes I am indeed using the split formulation by KKSACBulkC, KKSACBulkF, ACInterface, KKSSplitCHCRes, and SplitCHWRes. I am wondering why in the KKSACBulkC and KKSACBulkF, the residual and on-diagonal Jacobian terms are not multiplied by the Allen-Cahn mobility L, and only the off-diagonal Jacobian is multiplied by L?

**Reply by dschwen** [reply]
Ah, I see, the `L` is further up in the class hierarchy at

https://github.com/idaholab/moose/blob/7a8db3f1289a439b996e20314c310e7a1a833fe0/modules/phase_field/include/kernels/ACBulk.h#L84-L93

**Reply by dschwen** [reply]
Just follow the call chain. `computeDFDOP` is what gets overridden by the KKSACBulkC and KKSACBulkF kernels. The offdiagonal Jacobian is handled a bit differently, as it does not use anything from the ACBulk base class, and all contributions (including the `L`) are implemented there directly.

**Reply by xueyang94** [reply]
@dschwen Got it. With the local solve approach though, currently the KKS kernels directly inherit from the Kernel class. They just use the computeQpResidual and computeQpJacobian instead of computeDFDOP. In that case, the mobility L needs to be multiplied in the residual and Jacobian. The hierarchy could be changed in the future if needed. Thanks for clearing that up.
